### PR TITLE
Fix TypeWithAnnotations.ToTypeWithState() for (untyped) null literal

### LINKED ISSUE #46342: "CS8603: Possible null reference return" reported for null from lambda that returns T? where T : class
The compiler is inferring the lambda below returns `T` rather than `T?`, resulting in a warning for `return null;`.

This is a regression in behavior from the change to support `T?` for unconstrained `T` in https://github.com/dotnet/roslyn/pull/45993.
```C#
#nullable enable
using System;
class Program
{
    static void F<T>(Func<T> f)
    {
    }
    static void M<T>(bool b, T t) where T : class
    {
        F(() =>
        {
            if (b) return null; // warning CS8603: Possible null reference return
            return t;
        });
    }
}
```

### PR DESCRIPTION
**Pull request by cston**
Fixes #46342

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -51082,12 +51082,12 @@ static void G(string x)
...
-                // (9,54): warning CS8603: Possible null reference return.
+                // (9,9): warning CS8602: Dereference of a possibly null reference.
                 //         F(() => { if (x.Length > 0) return x; return null; }).ToString();
-                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "null").WithLocation(9, 54),
-                // (10,45): warning CS8603: Possible null reference return.
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(() => { if (x.Length > 0) return x; return null; })").WithLocation(9, 9),
+                // (10,9): warning CS8602: Dereference of a possibly null reference.
                 //         F(() => { if (x.Length == 0) return null; return x; }).ToString();
-                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "null").WithLocation(10, 45));
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(() => { if (x.Length == 0) return null; return x; })").WithLocation(10, 9));
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:51090]: Restores behavior from before https://github.com/dotnet/roslyn/pull/45993.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -133331,6 +133331,160 @@ internal interface IB : IA { }
...
+            await Task.Yield();
+            if (t is null) return null;
+            return t;
+        });
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyEmitDiagnostics();
+        }
+
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:133487]: Could we add similar tests with `default`? File an issue if misbehaving.

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:133487]: `return default;` results in "warning CS8603: Possible null reference return" with 16.7 and this branch.
`return default(T);` compiles without warnings with both. 

I'll add tests for those cases in a separate PR.

**Comment by RikkiGibson** [comment]
This needs to merge to master. We are reverting the feature in preview1 (see #46358).

**Pull request merged**